README.txt
* https://rubygems.org/gems/geohex_v3

## GEOHEX DESCRIPTION
GeoHex Gem for Ruby (c) 2009 Haruyuki Seki
http://twitter.com/hal_sk_e  (en)
http://twitter.com/hal_sk  (ja)

  * The GeoHex is a latitude/longitude encoding system invented by sa2da.
  * GeoHex divides geometry regions into hexagonal grid and set an unique identifier to each grid.
  * The coding idea is opened under the Creative Commons Attribution 2.1 Japan Lisence. 
  * GeoHex Documentation: https://sites.google.com/site/geohexdocs/
  * GeoHex Original Wiki(Japanese): http://geogames.net/labs/geohex
  * GeoHex Demo: http://geohex.net/v3.html
  * Various GeoHex Libraries: http://github.com/geohex/
  
## GEOHEX GEM DESCRIPTION
   
  This GeoHex Ruby gem can convert latitude/longitude to GeoHex code each others.
  * Encode from latitude/longitude to GeoHex code to an arbitrary level of precision
  * Decode from GeoHex code to latitude/longitude and level of precision
  * Return Zone from X,Y + Level
  * Return XY from L,L + level
  * Return XY From Code
  * Return Zone from Code
  

## INSTALL

   gem install geohex_v3

## QUICK START

   require 'geohex'
   GeoHex::Zone.encode(35.647401,139.716911,1)
   => '132KpuG' 
   
   GeoHex::Zone.decode('0dMV')
   => [24.338279000000004,124.1577708779443,7]
   
   GeoHex::Zone.getZoneByLocation(33.127120,-117.3274073, 11)
   => #<GeoHex::Zone:0x007fdbf180aca8
         @code="PC22751337146",
         @lat=33.127103696872936,
         @lon=-117.32741734265892,
         @x=-250028,
         @y=789182>
    
   GeoHex::Zone.getZoneByCode('PC22751337146')
   => #<GeoHex::Zone:0x007fce3d8758d0
         @code="PC22751337146",
         @lat=33.127103696872936,
         @lon=-117.32741734265892,
         @x=-250028,
         @y=789182>


## LICENSE

NOTE: 
The idea of GeoHex is licensed by sa2da. 
You should follow original license of GeoHex.
(The license is opened by CC-BY-SA license)
You must attribute this work to sa2da (with link).
http://geogames.net

This ruby code is covered by MIT License.
(The MIT License)

Copyright (c) 2009 Haruyuki SEKI 

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 3.0.0 / 2016-11-05

* will update these with stuff, ya know 



spec/geohexv3.js
/// Copyright (c) 2009 @sa2da (http://twitter.com/sa2da) ///
/// GeoHex is licensed under MIT License ///
/// http://www.geohex.org ///

///////////[2010.12.14 v3公開]/////////
///////////[2010.12.28 zoneByCode()内のlonを±180以内に補正]/////////
///////////[2011.9.11 zoneByCode()内のh_x,h_yを補正]/////////
///////////[2013.1.6 180度線をまたぐ場合のHEXのX,Y値を一意に補正]/////////
///////////[2013.1.6 範囲外のX,Y値を範囲内に補正する関数 adjustXY() を追加]/////////
///////////[2013.1.6 X,Y値からHEXを取得する関数 getZoneByXY() を追加]/////////
///////////[2013.1.6 NamespaceをGeoHexからGEOHEXに変更]/////////
///////////[2013.1.9 adjustXYの補正ロジックを修正]/////////
///////////[2013.1.11 getZoneByLocation()→getXYByLocation()とgetZoneByXY()に分割]/////////
///////////[2013.1.11 getZoneByCode()→getXYByCode()とgetZoneByXY()に分割]/////////
///////////[2013.1.11 adjustXY補正のx<yをx>yの場合に-180補正へ修正]/////////
///////////[2013.1.17 getZoneByXY()内 -180度補正Code生成ロジックを修正]/////////

(function (win) {

// namspace GEOHEX;
if (!win.GEOHEX)	win.GEOHEX = function(){};
// version: 3.2
GEOHEX.version = "3.2";
GEOHEX.cache_on = true;

// *** Share with all instances ***
var h_key = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
var h_base = 20037508.34;
var h_deg = Math.PI*(30/180);
var h_k = Math.tan(h_deg);

// private static
var _zoneCache = {};

// *** Share with all instances ***
// private static
function calcHexSize(_level) {
	return h_base/Math.pow(3, _level+3);
}

// private class
function Zone(lat, lon, x, y, code) {
	this.lat = lat;
	this.lon = lon;
	this.x = x;
	this.y = y;
	this.code = code;
}
Zone.prototype.getLevel = function () {
	return this.code.length-2;
};
Zone.prototype.getHexSize = function () {
	return calcHexSize(this.getLevel());
};
Zone.prototype.getHexCoords = function () {
	var h_lat = this.lat;
	var h_lon = this.lon;
	var h_xy = loc2xy(h_lon, h_lat);
	var h_x = h_xy.x;
	var h_y = h_xy.y;
	var h_deg = Math.tan(Math.PI * (60 / 180));
	var h_size = this.getHexSize();
	var h_top = xy2loc(h_x, h_y + h_deg *  h_size).lat;
	var h_btm = xy2loc(h_x, h_y - h_deg *  h_size).lat;

	var h_l = xy2loc(h_x - 2 * h_size, h_y).lon;
	var h_r = xy2loc(h_x + 2 * h_size, h_y).lon;
	var h_cl = xy2loc(h_x - 1 * h_size, h_y).lon;
	var h_cr = xy2loc(h_x + 1 * h_size, h_y).lon;
	return [
		{lat: h_lat, lon: h_l},
		{lat: h_top, lon: h_cl},
		{lat: h_top, lon: h_cr},
		{lat: h_lat, lon: h_r},
		{lat: h_btm, lon: h_cr},
		{lat: h_btm, lon: h_cl}
	];
};


function getZoneByLocation(_lat, _lon, _level) {
	var xy = getXYByLocation(_lat, _lon, _level);
	var zone = getZoneByXY(xy.x, xy.y, _level);
	return zone;
}

function getZoneByCode(_code) {
	var xy = getXYByCode(_code);
	var level = _code.length - 2;
	var zone = getZoneByXY(xy.x, xy.y, level);
	return zone;
}

function getXYByLocation(lat, lon, _level) {
	var h_size = calcHexSize(_level);
	var z_xy = loc2xy(lon, lat);
	var lon_grid = z_xy.x;
	var lat_grid = z_xy.y;
	var unit_x = 6 * h_size;
	var unit_y = 6 * h_size * h_k;
	var h_pos_x = (lon_grid + lat_grid / h_k) / unit_x;
	var h_pos_y = (lat_grid - h_k * lon_grid) / unit_y;
	var h_x_0 = Math.floor(h_pos_x);
	var h_y_0 = Math.floor(h_pos_y);
	var h_x_q = h_pos_x - h_x_0; 
	var h_y_q = h_pos_y - h_y_0;
	var h_x = Math.round(h_pos_x);
	var h_y = Math.round(h_pos_y);
	
	if (h_y_q > -h_x_q + 1) {
		if((h_y_q < 2 * h_x_q) && (h_y_q > 0.5 * h_x_q)){
			h_x = h_x_0 + 1;
			h_y = h_y_0 + 1;
		}
	} else if (h_y_q < -h_x_q + 1) {
		if ((h_y_q > (2 * h_x_q) - 1) && (h_y_q < (0.5 * h_x_q) + 0.5)){
			h_x = h_x_0;
			h_y = h_y_0;
		}
	}
	
	var inner_xy = adjustXY(h_x,h_y,_level);
	h_x = inner_xy.x;
	h_y = inner_xy.y;
	return {"x": h_x, "y":h_y};
}


function getXYByCode(code) {
	var level = code.length -2;
	var h_size =  calcHexSize(level);
	var unit_x = 6 * h_size;
	var unit_y = 6 * h_size * h_k;
	var h_x = 0;
	var h_y = 0;
	var h_dec9 =""+ (h_key.indexOf(code.charAt(0))*30+h_key.indexOf(code.charAt(1)))+code.substring(2);
	if(h_dec9.charAt(0).match(/[15]/)&&h_dec9.charAt(1).match(/[^125]/)&&h_dec9.charAt(2).match(/[^125]/)){
	  if(h_dec9.charAt(0)==5){
		h_dec9 = "7"+h_dec9.substring(1,h_dec9.length);
	  }else if(h_dec9.charAt(0)==1){
		h_dec9 = "3"+h_dec9.substring(1,h_dec9.length);
	  }
	}
	var d9xlen = h_dec9.length;
	for(i=0;i<level +3 - d9xlen;i++){
	  h_dec9 ="0"+h_dec9;
	  d9xlen++;
	}
	var h_dec3 = "";
	for(i=0;i<d9xlen;i++){
	  var h_dec0=parseInt(h_dec9.charAt(i)).toString(3);
	  if(!h_dec0){
	    h_dec3 += "00";
	  }else if(h_dec0.length==1){
	    h_dec3 += "0";
	  }
	  h_dec3 += h_dec0;
	}

	h_decx =new Array();
	h_decy =new Array();
	
	for(i=0;i<h_dec3.length/2;i++){
	  h_decx[i]=h_dec3.charAt(i*2);
	  h_decy[i]=h_dec3.charAt(i*2+1);
	}

	for(i=0;i<=level+2;i++){
	    var h_pow = Math.pow(3,level+2-i);
	    if(h_decx[i] == 0){
	        h_x -= h_pow;
	    }else if(h_decx[i] == 2){
	        h_x += h_pow;
	    }
	    if(h_decy[i] == 0){
	        h_y -= h_pow;
	    }else if(h_decy[i] == 2){
	        h_y += h_pow;
	    }
	}
	
	var inner_xy = adjustXY(h_x,h_y,level);
	h_x = inner_xy.x;
	h_y = inner_xy.y;
	
	return {"x":h_x, "y":h_y};
}

function getZoneByXY(_x, _y, _level) {
	var h_size = calcHexSize(_level);
	
	var h_x =_x;
	var h_y=_y;

	var unit_x = 6 * h_size;
	var unit_y = 6 * h_size * h_k;

	var h_lat = (h_k * h_x * unit_x + h_y * unit_y) / 2;
	var h_lon = (h_lat - h_y * unit_y) / h_k;

	var z_loc = xy2loc(h_lon, h_lat);
	var z_loc_x = z_loc.lon;
	var z_loc_y = z_loc.lat;
	
	var max_hsteps = Math.pow(3,_level+2);
	var hsteps = Math.abs(h_x - h_y);
	
	if(hsteps==max_hsteps){
		if(h_x>h_y){
		var tmp = h_x;
		h_x = h_y;
		h_y = tmp;
		}
		z_loc_x = -180;
	}
	
	var h_code ="";
	var code3_x =new Array();
	var code3_y =new Array();
	var code3 ="";
	var code9="";
	var mod_x = h_x;
	var mod_y = h_y;


	for(i = 0;i <= _level+2 ; i++){
	  var h_pow = Math.pow(3,_level+2-i);
	  if(mod_x >= Math.ceil(h_pow/2)){
	    code3_x[i] =2;
	    mod_x -= h_pow;
	  }else if(mod_x <= -Math.ceil(h_pow/2)){
	    code3_x[i] =0;
	    mod_x += h_pow;
	  }else{
	    code3_x[i] =1;
	  }
	  if(mod_y >= Math.ceil(h_pow/2)){
	    code3_y[i] =2;
	    mod_y -= h_pow;
	  }else if(mod_y <= -Math.ceil(h_pow/2)){
	    code3_y[i] =0;
	    mod_y += h_pow;
	  }else{
	    code3_y[i] =1;
	  }
	  if(i==2&&(z_loc_x==-180 || z_loc_x>=0)){
		  if(code3_x[0]==2&&code3_y[0]==1&&code3_x[1]==code3_y[1]&&code3_x[2]==code3_y[2]){
			code3_x[0]=1;
			code3_y[0]=2;
		  }else if(code3_x[0]==1&&code3_y[0]==0&&code3_x[1]==code3_y[1]&&code3_x[2]==code3_y[2]){
			code3_x[0]=0;
			code3_y[0]=1
		  }
	  }
	}

	for(i=0;i<code3_x.length;i++){
	  code3 += ("" + code3_x[i] + code3_y[i]);
	  code9 += parseInt(code3,3);
	  h_code += code9;
	  code9="";
	  code3="";
	}
	var h_2 = h_code.substring(3);
	var h_1 = h_code.substring(0,3);
	var h_a1 = Math.floor(h_1/30);
	var h_a2 = h_1%30;
	h_code = (h_key.charAt(h_a1)+h_key.charAt(h_a2)) + h_2;

	if(GEOHEX.cache_on){
		if (!!_zoneCache[h_code])	return _zoneCache[h_code];
		return (_zoneCache[h_code] = new Zone(z_loc_y, z_loc_x, _x, _y, h_code));
	}else{
		return new Zone(z_loc_y, z_loc_x, _x, _y, h_code);
	}
}

function adjustXY(_x, _y, _level){
	var x =_x;
	var y =_y;
	var rev = 0;
	var max_hsteps = Math.pow(3,_level+2);
	var hsteps = Math.abs(x - y);
	if(hsteps==max_hsteps&&x>y){
		var tmp = x;
		x = y;
		y = tmp;
		rev =1;
	}else if(hsteps>max_hsteps){
		var dif = hsteps - max_hsteps;
		var dif_x = Math.floor(dif/2);
		var dif_y = dif - dif_x;
		var edge_x;
		var edge_y;
		if(x>y){
			edge_x = x - dif_x;
			edge_y = y + dif_y;
			var h_xy = edge_x;
			edge_x = edge_y;
			edge_y = h_xy;
			x = edge_x + dif_x;
			y = edge_y - dif_y;
		}else if(y>x){
			edge_x = x + dif_x;
			edge_y = y - dif_y;
			var h_xy = edge_x;
			edge_x = edge_y;
			edge_y = h_xy;
			x = edge_x - dif_x;
			y = edge_y + dif_y;
		}
	}
	return { x: x, y: y , rev:rev};
}

// private static
function loc2xy(lon, lat) {
	var x = lon * h_base / 180;
	var y = Math.log(Math.tan((90 + lat) * Math.PI / 360)) / (Math.PI / 180);
	y *= h_base / 180;
	return { x: x, y: y };
}
// private static
function xy2loc(x, y) {
	var lon = (x / h_base) * 180;
	var lat = (y / h_base) * 180;
	lat = 180 / Math.PI * (2 * Math.atan(Math.exp(lat * Math.PI / 180)) - Math.PI / 2);
	return { lon: lon, lat: lat };
}

// EXPORT
GEOHEX.getZoneByLocation = getZoneByLocation;
GEOHEX.getXYByLocation = getXYByLocation;
GEOHEX.getZoneByCode = getZoneByCode;
GEOHEX.getZoneByXY = getZoneByXY;
GEOHEX.loc2xy = loc2xy;
GEOHEX.xy2loc = xy2loc;
GEOHEX.adjustXY = adjustXY;

})(this);

lib/geohex.rb
# -*- coding: utf-8 -*-
#
# ported from js libraries http://geohex.net
# author Haruyuki Seki 
#
require 'rubygems'
require 'ostruct'

module GeoHex
  VERSION = '3.0.1'

  H_KEY = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz" 
  H_BASE = 20037508.34 
  H_DEG = Math::PI*(30.0/180) 
  H_K = Math.tan(H_DEG) 
  R = 20037508.34
  
  class Zone
    attr_accessor :code, :lat, :lon, :x, :y
    def initialize(*params)
      raise ArgumentError, "latlng or code is needed" if params.size == 0

      @lat = params[0]
      @lon = params[1]
      @x   = params[2]
      @y   = params[3]
      @code= params[4]
      
    end
    
    def self.calcHexSize(level)
      H_BASE/(3**(level+3))
    end
    
    def level
      H_KEY.index(code[0,1])
    end
    def hexSize
      self.calcHexSize(self.level)
    end
    
    def hexCoords
      h_lat = self.lat 
      h_lon = self.lon 
      h_xy = GeoHex::Zone.loc2xy(h_lon, h_lat) 
      h_x = h_xy.x
      h_y = h_xy.y
      h_deg = Math.tan(Math::PI * (60 / 180)) 
      h_size = self.hexSize
      h_top = xy2loc(h_x, h_y + h_deg *  h_size).lat 
      h_btm = xy2loc(h_x, h_y - h_deg *  h_size).lat 
      
      h_l = xy2loc(h_x - 2 * h_size, h_y).lon 
      h_r = xy2loc(h_x + 2 * h_size, h_y).lon 
      h_cl = xy2loc(h_x - 1 * h_size, h_y).lon 
      h_cr = xy2loc(h_x + 1 * h_size, h_y).lon 
      
      [
       {:lat => h_lat, :lon => h_l},
       {:lat => h_top, :lon => h_cl},
       {:lat => h_top, :lon => h_cr},
       {:lat => h_lat, :lon => h_r},
       {:lat => h_btm, :lon => h_cr},
       {:lat => h_btm, :lon => h_cl}
      ]
    end
    
    # latlon to geohex
    def self.encode(lat,lon,level=7)
      raise ArgumentError, "latitude must be double" unless (lat.is_a?(Numeric))
      raise ArgumentError, "latitude must be between -85 and 85" if (lat < -85.1 || lat > 85.1)
      raise ArgumentError, "latitude must be between -85 and 85" if (lat < -85.1 || lat > 85.1)
      raise ArgumentError, "longitude must be between -180 and 180" if (lon < -180 || lon > 180)
      raise ArgumentError, "level must be between 0 and 24" if (level < 0 || level > 24)
      
      h_size = self.calcHexSize(level)
      z_xy = loc2xy(lon,lat)
      lon_grid = z_xy.x
      lat_grid = z_xy.y
      unit_x = 6.0* h_size 
      unit_y = 6.0* h_size*H_K 
      h_pos_x = (lon_grid + lat_grid / H_K) / unit_x 
      h_pos_y = (lat_grid - H_K*lon_grid) / unit_y 

      h_x_0 = h_pos_x.floor 
      h_y_0 = h_pos_y.floor 
      h_x_q = h_pos_x - h_x_0
      h_y_q = h_pos_y - h_y_0
      h_x = h_pos_x.round 
      h_y = h_pos_y.round 
      
      h_max=(H_BASE/unit_x + H_BASE/unit_y).round 
      
      if ( h_y_q > -h_x_q + 1 ) 
        if ( h_y_q < (2 * h_x_q ) and  h_y_q > (0.5 * h_x_q ) )
          h_x = h_x_0 + 1
          h_y = h_y_0 + 1
        end
      elsif ( h_y_q < -h_x_q + 1 ) 
        if ( (h_y_q > (2 * h_x_q ) - 1 ) && ( h_y_q < ( 0.5 * h_x_q ) + 0.5 ) ) 
          h_x = h_x_0
          h_y = h_y_0
        end
      end
    
      h_lat = (H_K * h_x * unit_x + h_y * unit_y) / 2 
      h_lon = (h_lat - h_y * unit_y)/H_K 
      
      z_loc = xy2loc(h_lon,h_lat)
      z_loc_x = z_loc.lon
      z_loc_y = z_loc.lat
      
      if (H_BASE - h_lon <h_size)
        z_loc_x = 180.0 
        h_xy = h_x 
        h_x = h_y 
        h_y = h_xy 
      end
      
      h_x_p = (h_x<0) ? 1 : 0 
      h_y_p = (h_y<0) ? 1 : 0 
      h_x_abs = ((h_x).abs * 2 + h_x_p).to_f 
      h_y_abs = ((h_y).abs * 2 + h_y_p).to_f 
      h_x_10000 = ((h_x_abs%777600000)/12960000).floor 
      h_x_1000 = ((h_x_abs%12960000)/216000).floor 
      h_x_100 = ((h_x_abs%216000)/3600).floor 
      h_x_10 = ((h_x_abs%3600)/60).floor 
      h_x_1 = ((h_x_abs%3600)%60).floor 
      h_y_10000 = ((h_y_abs%77600000)/12960000).floor 
      h_y_1000 = ((h_y_abs%12960000)/216000).floor 
      h_y_100 = ((h_y_abs%216000)/3600).floor 
      h_y_10 = ((h_y_abs%3600)/60).floor 
      h_y_1 = ((h_y_abs%3600)%60).floor 
      h_code = H_KEY[level % 60, 1] 
      h_code += H_KEY[h_x_10000, 1]+H_KEY[h_y_10000, 1] if(h_max >=12960000/2)  
      h_code += H_KEY[h_x_1000, 1]+H_KEY[h_y_1000, 1] if(h_max >=216000/2)  
      h_code += H_KEY[h_x_100, 1]+H_KEY[h_y_100, 1] if(h_max >=3600/2)  
      h_code += H_KEY[h_x_10, 1]+H_KEY[h_y_10, 1] if(h_max >=60/2)  
      h_code += H_KEY[h_x_1, 1]+H_KEY[h_y_1, 1] 
      
      return h_code 
    end
  
    # geohex to latlon
    def self.decode(code)
      c_length = code.length 
      level = H_KEY.index(code[0,1]) 
      scl = level 
      h_size =  self.calcHexSize(level) 
      unit_x = 6.0 * h_size 
      unit_y = 6.0 * h_size * H_K 
      h_max = (H_BASE / unit_x + H_BASE / unit_y).round 
      h_x = 0 
      h_y = 0 

      if (h_max >= 12960000 / 2) 
        h_x = H_KEY.index(code[1,1]) * 12960000 + 
          H_KEY.index(code[3,1]) * 216000 + 
          H_KEY.index(code[5,1]) * 3600 + 
          H_KEY.index(code[7,1]) * 60 + 
          H_KEY.index(code[9,1]) 
        h_y = H_KEY.index(code[2,1]) * 12960000 + 
          H_KEY.index(code[4,1]) * 216000 + 
          H_KEY.index(code[6,1]) * 3600 + 
          H_KEY.index(code[8,1]) * 60 + 
          H_KEY.index(code[10,1]) 
      elsif (h_max >= 216000 / 2) 
        h_x = H_KEY.index(code[1,1]) * 216000 + 
          H_KEY.index(code[3,1]) * 3600 + 
          H_KEY.index(code[5,1]) * 60 + 
          H_KEY.index(code[7,1]) 
        h_y = H_KEY.index(code[2,1]) * 216000 + 
          H_KEY.index(code[4,1]) * 3600 + 
          H_KEY.index(code[6,1]) * 60 + 
          H_KEY.index(code[8,1]) 
      elsif (h_max >= 3600 / 2) 
        h_x = H_KEY.index(code[1,1]) * 3600 + 
          H_KEY.index(code[3,1]) * 60 + 
          H_KEY.index(code[5,1]) 
		h_y = H_KEY.index(code[2,1]) * 3600 + 
          H_KEY.index(code[4,1]) * 60 + 
          H_KEY.index(code[6,1]) 
      elsif (h_max >= 60 / 2) 
        h_x = H_KEY.index(code[1,1]) * 60 + 
          H_KEY.index(code[3,1]) 
        h_y = H_KEY.index(code[2,1]) * 60 + 
          H_KEY.index(code[4,1]) 
      else
        h_x = H_KEY.index(code[1,1]) 
        h_y = H_KEY.index(code[2,1]) 
      end
      h_x = (h_x % 2 == 1) ? -(h_x - 1) / 2 : h_x / 2 
      h_y = (h_y % 2 == 1) ? -(h_y - 1) / 2 : h_y / 2 

      h_lat_y = (H_K * h_x * unit_x + h_y * unit_y) / 2 
      h_lon_x = (h_lat_y - h_y * unit_y) / H_K 

      h_loc = xy2loc(h_lon_x, h_lat_y) 
      return [h_loc.lat, h_loc.lon, level]
    end
    
    def self.getZoneByXy(_x,_y, _level)

    	h_size = self.calcHexSize(_level) 
	
    	h_x = _x
    	h_y = _y
      
    	unit_x = 6 * h_size
    	unit_y = 6 * h_size * H_K

    	h_lat = (H_K * h_x * unit_x + h_y * unit_y) / 2
    	h_lon = (h_lat - h_y * unit_y) / H_K

    	z_loc = self.xy2loc(h_lon, h_lat)
    	z_loc_x = z_loc.lon
    	z_loc_y = z_loc.lat
      
    	max_hsteps = 3**(_level + 2)
    	hsteps = (h_x - h_y).abs

    	if hsteps == max_hsteps
    		if h_x > h_y
      		tmp = h_x
      		h_x = h_y
      		h_y = tmp
        end
    		z_loc_x = -180
      end
      
    	h_code = ""
       
    	code3_x = Array.new
    	code3_y = Array.new
    	code3 = ""
    	code9 = ""
    	mod_x = h_x
    	mod_y = h_y
    
    	(0..(_level + 2)).each do |i| 
    	  h_pow = 3**(_level + 2 - i)
        
    	  if mod_x >= ((h_pow.to_f)/2).ceil 
    	    code3_x[i] =2 
    	    mod_x -= h_pow 
        elsif mod_x <= -((h_pow.to_f)/2).ceil 
    	    code3_x[i] =0 
    	    mod_x += h_pow 
        else
    	    code3_x[i] =1 
        end
        

        
    	  if mod_y >= ((h_pow.to_f)/2).ceil
    	    code3_y[i] = 2 
    	    mod_y -= h_pow 
        elsif mod_y <= -((h_pow.to_f)/2).ceil
    	    code3_y[i] = 0 
    	    mod_y += h_pow 
    	  else
    	    code3_y[i] = 1 
        end
      

        
    	  if i==2 && (z_loc_x==-180 || z_loc_x>=0) 
    		  if code3_x[0] == 2 && code3_y[0] == 1 && code3_x[1] == code3_y[1] && code3_x[2] == code3_y[2] 
      			code3_x[0] = 1
      			code3_y[0] = 2
          elsif code3_x[0] == 1 && code3_y[0] == 0 && code3_x[1] == code3_y[1] && code3_x[2] == code3_y[2] 
      			code3_x[0] = 0
      			code3_y[0] = 1
          end
        end
      end
         
    	code3_x.each_with_index do |x,i|
    	  code3  += "#{code3_x[i]}#{code3_y[i]}"
    	  code9  += code3.to_i(3).to_s
    	  h_code += code9
    	  code9 = ""
    	  code3 = ""
      end
      
    	h_2 = h_code[3..h_code.length]
    	h_1 = h_code[0..2]
    	h_a1 = (h_1.to_f/30).floor
    	h_a2 = h_1.to_i%30
      
        
    	h_code ="#{H_KEY[h_a1]}#{H_KEY[h_a2]}#{h_2}"


     return GeoHex::Zone.new(z_loc_y, z_loc_x, _x, _y, h_code) 
      
    end
    
    def self.getXYByLocation(lat, lon, _level)
    	h_size = calcHexSize(_level)
    	z_xy = loc2xy(lon, lat)
    	lon_grid = z_xy.x
    	lat_grid = z_xy.y
    	unit_x = 6 * h_size
    	unit_y = 6 * h_size * H_K
    	h_pos_x = (lon_grid + lat_grid / H_K) / unit_x
    	h_pos_y = (lat_grid - H_K * lon_grid) / unit_y
    	h_x_0 = (h_pos_x).floor
    	h_y_0 = (h_pos_y).floor
    	h_x_q = h_pos_x - h_x_0;
    	h_y_q = h_pos_y - h_y_0
    	h_x = (h_pos_x).round
    	h_y = (h_pos_y).round
	
    	if h_y_q > -h_x_q + 1
    		if h_y_q < 2 * h_x_q && h_y_q > 0.5 * h_x_q
    			h_x = h_x_0 + 1
    			h_y = h_y_0 + 1
        end
      elsif h_y_q < -h_x_q + 1
    		if h_y_q > (2 * h_x_q) - 1 && h_y_q < 0.5 * h_x_q + 0.5
    			h_x = h_x_0
    			h_y = h_y_0
        end
      end
	
    	inner_xy = adjustXY(h_x, h_y, _level)
    	h_x = inner_xy[:x]
    	h_y = inner_xy[:y]
    	
      {
        x: h_x, 
        y: h_y 
      };
    end
    
    def self.getZoneByLocation(_lat, _lon, _level) 
    	xy = self.getXYByLocation(_lat, _lon, _level)
    	zone = self.getZoneByXy(xy[:x], xy[:y], _level)
    	
      zone
    end
    
    def self.getXYByCode(code)
    	level = code.length - 2
    	h_size =  calcHexSize(level)
    	unit_x = 6 * h_size
    	unit_y = 6 * h_size * H_K
    	h_x = 0
    	h_y = 0

    	h_dec9 = (( H_KEY.index(code[0]) * 30 + H_KEY.index(code[1])).to_s + code[2..-1].to_s)
      
    	if h_dec9[0].match(/[15]/) && h_dec9[1].match(/[^125]/) && h_dec9[2].match(/[^125]/)
    	  if h_dec9[0] == 5
    		  h_dec9 = "7" + h_dec9[1..(h_dec9.length-1)]
        elsif h_dec9[0] == 1
    		  h_dec9 = "3" + h_dec9[1..h_dec9.length-1]
        end
      end

      d9xlen = h_dec9.length
      state = (level + 3 - d9xlen)
      if state != 0 
        (0..state-1).each do |i|
          h_dec9 = "0" + h_dec9
          d9xlen += 1
        end
      end
      
    	h_dec3 = String.new()
    	      
      (0..(d9xlen-1)).each do |i|
    	  h_dec0 = h_dec9[i].to_i.to_s(3)

    	  if !h_dec0 
    	    h_dec3 += "00"
        elsif h_dec0.length == 1
    	    h_dec3 += "0"
        end
    	  h_dec3 += h_dec0;
      end

    	h_decx =[]
    	h_decy =[]
	
      (0..((h_dec3.length/2)-1)).each do |i|
    	  x = i*2
        y = i*2+1
        h_decx[i] = h_dec3[x]
    	  h_decy[i] = h_dec3[y]
      end

    	(0..level+2).each do |i|
  	    h_pow = 3**(level+2-i)

  	    if h_decx[i].to_i == 0
  	        h_x -= h_pow;
  	    elsif h_decx[i].to_i == 2
  	        h_x += h_pow;
        end
        
        if h_decy[i].to_i == 0
  	        h_y -= h_pow
  	    elsif h_decy[i].to_i == 2
  	        h_y += h_pow
        end
      
      end


      
    	inner_xy = adjustXY(h_x,h_y,level)
    	h_x = inner_xy[:x]
    	h_y = inner_xy[:y]
	
    	{
        x: h_x, 
        y: h_y
      }
    end
    
    def self.getZoneByCode(code)
    	xy = getXYByCode(code)
    	level = code.length - 2
      zone = getZoneByXy(xy[:x], xy[:y], level)
      
      zone
    end
    
  end
    
  class << Zone
    
    def loc2xy(_lon,_lat) 
      x=_lon*H_BASE/180 
      y= Math.log(Math.tan((90+_lat)*Math::PI/360)) / (Math::PI / 180 ) 
      y= y * H_BASE / 180 
      return OpenStruct.new("x" => x, "y" => y) 
    end
    
    def xy2loc(_x,_y) 
      lon=(_x/H_BASE)*180 
      lat=(_y/H_BASE)*180 
    lat=180.0/Math::PI*(2.0*Math.atan(Math.exp(lat*Math::PI/180))-Math::PI/2) 
      return OpenStruct.new("lon" => lon,"lat" => lat) 
    end
    
    def adjustXY(_x, _y, _level)
    	x = _x
    	y = _y
    	rev = 0
    	max_hsteps = 3**(_level+2)
    	hsteps = (x - y).abs
    	if hsteps == max_hsteps && x>y
    		tmp = x
    		x = y
    		y = tmp
    		rev = 1
      elsif hsteps > max_hsteps
    		dif = hsteps - max_hsteps
    		dif_x = (dif/2).floor
    		dif_y = dif - dif_x
    		if x > y 
    			edge_x = x - dif_x
    			edge_y = y + dif_y
    			h_xy = edge_x
    			edge_x = edge_y
    			edge_y = h_xy
    			x = edge_x + dif_x
    			y = edge_y - dif_y
        elsif y > x
    			edge_x = x + dif_x
    			edge_y = y - dif_y
    			h_xy = edge_x
    			edge_x = edge_y
    			edge_y = h_xy
    			x = edge_x - dif_x
    			y = edge_y + dif_y
        end
      end
    	
      { x: x, y: y, rev: rev }
    end
    
  end
end
